/* Copyright 2023 The FedLearner Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = "proto3";

package fedlearner_webconsole.proto.rpc.v2;

import "google/protobuf/empty.proto";
import "fedlearner_webconsole/proto/project.proto";
import "fedlearner_webconsole/proto/workflow_definition.proto";

message CreatePendingProjectRequest {
    string name = 1;
    string uuid = 2;
    ParticipantsInfo participants_info = 3;
    string comment = 4;
    // Username of creator in Coordinator
    string creator_username = 5;
    ProjectConfig config = 6;
    string ticket_uuid = 8;
}

message UpdatePendingProjectRequest {
    string uuid = 1;
    map<string, ParticipantInfo> participants_map = 2;
}

message SyncPendingProjectStateRequest {
    string uuid = 1;
    // ACCEPTED or CLOSED
    string state = 2;
}

message CreateProjectRequest {
    string uuid = 1;
}

message DeletePendingProjectRequest {
    string uuid = 1;
}

message SendTemplateRevisionRequest {
    WorkflowDefinition config = 1;
    string name = 2;
    string comment = 3;
    // WorkflowTemplateKind enum
    string kind = 4;
    int64 revision_index = 5;
}

service ProjectService {
    // Only Coordinator used method
    rpc CreatePendingProject (CreatePendingProjectRequest) returns (google.protobuf.Empty) {}
    // Only Coordinator used method
    rpc UpdatePendingProject (UpdatePendingProjectRequest) returns (google.protobuf.Empty) {}
    // Only Coordinator used method
    rpc DeletePendingProject (DeletePendingProjectRequest) returns (google.protobuf.Empty) {}
    // Only Coordinator used method
    rpc CreateProject (CreateProjectRequest) returns (google.protobuf.Empty) {}
    // Only Participant used method
    rpc SyncPendingProjectState (SyncPendingProjectStateRequest) returns (google.protobuf.Empty) {}

    rpc SendTemplateRevision (SendTemplateRevisionRequest) returns (google.protobuf.Empty) {}
}
